interface/ceed-preconditioning.c

```diff
@@ -518,6 +518,77 @@ static int CeedSingleOperatorAssembleSymbolic(CeedOperator op, CeedInt offset, C
+the same ordering.
+
+   This will generally be slow unless your operator is low-order.
+
+   Note: Calling this function asserts that setup is complete and sets the CeedOperator as immutable.
+
+   @param[in]  op          CeedOperator to assemble
+   @param[out] num_entries Number of entries in coordinate nonzero pattern
+   @param[out] rows        Row number for each entry
+   @param[out] cols        Column number for each entry
+
+   @ref User
+**/
+int CeedOperatorLinearAssemblePointBlockDiagonalSymbolic(CeedOperator op, CeedSize *num_entries, CeedInt **rows, CeedInt **cols) {
+  Ceed ceed;
+  bool is_composite;
+  CeedCall(CeedOperatorGetCeed(op, &ceed));
+  CeedCall(CeedOperatorIsComposite(op, &is_composite));
+
+  CeedSize input_size = 0, output_size = 0;
+  CeedCall(CeedOperatorGetActiveVectorLengths(op, &input_size, &output_size));
+  CeedCheck(input_size == output_size, ceed, CEED_ERROR_DIMENSION, "Operator must be square");
+
+  CeedSize num_output_components;
+  if (is_composite) {
+    CeedInt                  num_suboperators;
+    CeedOperator            *sub_operators;
+    CeedOperatorAssemblyData data;
+    CeedCall(CeedCompositeOperatorGetNumSub(op, &num_suboperators));
+    CeedCall(CeedCompositeOperatorGetSubList(op, &sub_operators));
+
+    CeedCall(CeedOperatorGetOperatorAssemblyData(sub_operators[0], &data));
+    CeedCall(CeedOperatorAssemblyDataGetEvalModes(data, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &num_output_components));
+    // Verify that all sub-operators have same num_output_components
+    for (CeedInt k = 1; k < num_suboperators; k++) {
+      CeedSize num_output_components_sub;
+      CeedCall(CeedOperatorGetOperatorAssemblyData(sub_operators[k], &data));
+      CeedCall(CeedOperatorAssemblyDataGetEvalModes(data, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &num_output_components_sub));
+      CeedCheck(num_output_components == num_output_components_sub, ceed, CEED_ERROR_INCOMPATIBLE,
+                "All suboperators must have the same number of output components");
+    }
+  } else {
+    CeedOperatorAssemblyData data;
+    CeedCall(CeedOperatorGetOperatorAssemblyData(op, &data));
+    CeedCall(CeedOperatorAssemblyDataGetEvalModes(data, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &num_output_components));
+  }
+  *num_entries = input_size * num_output_components;
+  CeedCall(CeedCalloc(*num_entries, rows));
+  CeedCall(CeedCalloc(*num_entries, cols));
+
+  CeedInt num_nodes = input_size / num_output_components;
+  for (CeedSize n = 0; n < num_nodes; n++) {
+    for (CeedInt i = 0; i < num_output_components; i++) {
+      for (CeedInt j = 0; j < num_output_components; j++) {
+        (*rows)[(n * num_output_components + i) * num_output_components + j] = n * num_output_components + i;
+        (*cols)[(n * num_output_components + i) * num_output_components + j] = n * num_output_components + j;
+      }
+    }
+  }
```

Comment:
Note that this may not be accurate for element restrictions with component stride != 1. Currently, the output of `CeedOperatorLinearAssemblePointBlockDiagonal` has components collapsed together, regardless of the component stride for the active element restriction.

I'm not sure if `CeedOperatorLinearAssemble` and `CeedOperatorLinearAssemble` does the same, or if it tried to assemble it into a matrix that would operate directly on the L-vector's ordering.

Reply:
We need it to either be calculated correctly for general stride or to error with an explanation that it's not currently supported (and we'll add it if/when there is demand).

Reply:
I think for right now, erroring out with an explanation is probably the way to go until I get time to do a full implementation.

Thinking outloud, wouldn't this also assume successive offsets as well? I guess I should probably check that `offsets[i+1] > offsets[i]` in addition to component stride length.

Reply:
I don't think we require positive stride, but do you mean `offsets[n+1] > offsets[n]` (index by node)? That needn't be guaranteed either.